pkg/extension/update_interceptor.go

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package extension
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Peripli/service-manager/pkg/query"
+
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/storage"
+)
+
+type namedUpdateAPIFunc struct {
+	Name string
+	Func func(InterceptUpdateOnAPI) InterceptUpdateOnAPI
+}
+
+type namedUpdateTxFunc struct {
+	Name string
+	Func func(InterceptUpdateOnTx) InterceptUpdateOnTx
+}
+
+type updateHookOnAPIHandler struct {
+	UpdateHookOnAPIFuncs         []*namedUpdateAPIFunc
+	UpdateHookOnTransactionFuncs []*namedUpdateTxFunc
+}
+
+func (c *updateHookOnAPIHandler) OnAPIUpdate(f InterceptUpdateOnAPI) InterceptUpdateOnAPI {
+	for i := range c.UpdateHookOnAPIFuncs {
+		f = c.UpdateHookOnAPIFuncs[len(c.UpdateHookOnAPIFuncs)-1-i].Func(f)
+	}
+	return f
+}
+
+func (c *updateHookOnAPIHandler) OnTxUpdate(f InterceptUpdateOnTx) InterceptUpdateOnTx {
+	for i := range c.UpdateHookOnTransactionFuncs {
+		f = c.UpdateHookOnTransactionFuncs[len(c.UpdateHookOnTransactionFuncs)-1-i].Func(f)
+	}
+	return f
+}
+
+// UnionUpdateInterceptor returns a function which spawns all update interceptors, sorts them and wraps them into one.
+func UnionUpdateInterceptor(providers []UpdateInterceptorProvider) func() UpdateInterceptor {
```

Comment:
I think this a bit different - this function accepts *provider and returns *interceptor.

Reply:
Do you mean something like this:

```go
type CreateInterceptorProviders []CreateInterceptorProvider

func (cip CreateInterceptorProviders) Provide() func() CreateInterceptor {
	return func() CreateInterceptor {
		c := &createHookOnAPIHandler{}
		c.CreateHookOnAPIFuncs = make([]*namedCreateAPIFunc, 0, len(cip))
		c.CreateHookOnTransactionFuncs = make([]*namedCreateTxFunc, 0, len(cip))

		for _, p := range cip {
			hook := p.Provide()
			positionAPIType := PositionNone
			positionTxType := PositionNone
			nameAPI := ""
			nameTx := ""

			if orderedProvider, isOrdered := p.(Ordered); isOrdered {
				positionAPIType, nameAPI = orderedProvider.PositionAPI()
				positionTxType, nameTx = orderedProvider.PositionTx()
			}

			c.insertAPIFunc(positionAPIType, nameAPI, &namedCreateAPIFunc{
				Name: p.Name(),
				Func: hook.AroundTxCreate,
			})
			c.insertTxFunc(positionTxType, nameTx, &namedCreateTxFunc{
				Name: p.Name(),
				Func: hook.OnTxCreate,
			})
		}
		return c
	}
}
```

and  usage:

```go
createInterceptorProvider := CreateInterceptorProviders(ir.createProviders[objectType]).Provide()
```

instead of:

```go
createInterceptor := UnionCreateInterceptor(ir.createProviders[objectType])()
```